src/extensions/agent_modules/cloud_communication/iothub_client/module/src/iothub_client_module.c

```diff
@@ -0,0 +1,694 @@
+        Log_Error("Failed to set DO connection string in Nested Edge scenario, result: 0x%08x", result.ResultCode);
+        goto done;
+    }
+
+    result.ResultCode = ADUC_GeneralResult_Success;
+    result.ExtendedResultCode = 0;
+
+done:
+
+    ADUC_ConnectionInfo_DeAlloc(&info);
+    return result.ResultCode == ADUC_GeneralResult_Success ? 0 : -1;
+}
+
+/**
+ * @brief
+ *
+ * @param module
+ *
+ * @return ADUC_Result
+ */
+int IoTHubClientModule_Deinitialize(ADUC_AGENT_MODULE_HANDLE module)
+{
+    Log_Info("Deinitialize");
+    ADUC_D2C_Messaging_Uninit();
+#ifdef ADUC_COMMAND_HELPER_H
+    UninitializeCommandListenerThread();
+#endif
+    ADUC_PnP_Components_Destroy();
+    IoTHub_CommunicationManager_Deinit();
+    DiagnosticsComponent_DestroyDeviceName();
+    ADUC_Logging_Uninit();
+    // TODO: nox-msft Unload all extension
+    return 0;
+}
+
+/**
+ * @brief Create a Device Update Agent Module for IoT Hub PnP Client.
+ *
+ * @return ADUC_AGENT_MODULE_HANDLE The handle to the module.
+ */
+ADUC_AGENT_MODULE_HANDLE IoTHubClientModule_Create()
+{
+    ADUC_AGENT_MODULE_HANDLE handle = &g_iotHubClientHandle;
+    return handle;
+}
+
+void IoTHubClientModule_Destroy(ADUC_AGENT_MODULE_HANDLE handle)
+{
+    IGNORED_PARAMETER(handle);
+}
+
+/**
+ * @brief Perform the work for the extension. This must be a non-blocking operation.
+ *
+ * @return ADUC_Result
+ */
+int IoTHubClientModule_DoWork(ADUC_AGENT_MODULE_HANDLE handle)
+{
+    if (handle != &g_iotHubClientHandle)
+    {
+        goto done;
+    }
+
+    // If any components have requested a DoWork callback, regularly call it.
+    for (unsigned index = 0; index < ARRAY_SIZE(componentList); ++index)
+    {
+        PnPComponentEntry* entry = componentList + index;
+
+        if (entry->DoWork != NULL)
+        {
+            entry->DoWork(entry->Context);
+        }
+    }
+
+    ADUC_D2C_Messaging_DoWork();
+
+    // NOTE: When using low level samples (iothub_ll_*), the IoTHubDeviceClient_LL_DoWork
+    // function must be called regularly (eg. every 100 milliseconds) for the IoT device client to work properly.
+    // See: https://github.com/Azure/azure-iot-sdk-c/tree/master/iothub_client/samples
+    // NOTE: For this example the above has been wrapped to support module and device client methods using
+    // the client_handle_helper.h function ClientHandle_DoWork()
+    IoTHub_CommunicationManager_DoWork(&g_iotHubClientHandle);
+
+done:
+    return 0;
+}
+
+/**
+ * @brief Get the Data object for the specified key.
+ *
+ * @param handle agent module handle
+ * @param dataType data type
+ * @param key data key/name
+ * @param buffer return buffer (call must free the memory of the return value once done with it)
+ * @param size return size of the return value
+ * @return int 0 on success
+*/
+int IoTHubClientModule_GetData(ADUC_AGENT_MODULE_HANDLE handle, ADUC_MODULE_DATA_TYPE dataType, const char* key, void** buffer, int* size)
+{
+    IGNORED_PARAMETER(handle);
```

Comment:
UNUSED_PARAMETER has been used up to now so seems it should use this instead of adding yet another one that does same thing.